src/object/obj_rpc.c

```diff
@@ -755,7 +755,7 @@ crt_proc_struct_daos_cpd_sub_req(crt_proc_t proc, crt_proc_op_t proc_op,
 				if (unlikely(rc))
 					D_GOTO(out, rc);
 			}
-		} else {
+		} else if (!(dcu->dcu_flags & ORF_EMPTY_SGL)) {
```

Comment:
if pack a "sgl_nr" for encoding which is 0 if "dcu_sgls == NULL", and for encoding unpack and check sgl_nr first, seems need not add ORF_EMPTY_SGL flag? (just to save one flag bit if needed).
BTW, is it needed to upgrade rpc version?

Reply:
If we add new filed on wire, then it will cause old server (2.2) crash. We have to consider the interoperability issue. Since we will fix it on 2.4, which version (9) is already higher than 2.2 (8), it is unnecessary to bump the version again. Anyway, I need to refresh the patch to refuse new client (2.4) to send the CPD RPC with empty SGL and non-zero nr to old server (2.2), otherwise, in spite of how we pack on wire, the old server will always fail to handle that.

Reply:
If we only fix the client code to pack `nr` empty `sgls` in release/2.2 and master, but make no other changes to the protocol and the engine code, will we be fine? The idea is that this is a bug in the existing client code, and we don't attempt to make engines work with buggy clients. If a user has an existing 2.2 system, he suffers from this bug. If he upgrades the server side to 2.4, he still suffers from the bug, rightfully. To address the bug, he needs to upgrade the client side to a newer 2.2 or newer 2.4 version. Then we avoid introducing interoperability issues here.

Reply:
Similarly, we could also ask if only fixing the engine code to unpack such problematic messages properly, will we be fine? A similar line of argument also avoids introducing interoperability issues here.

Reply:
> The idea is that this is a bug in the existing client code, and we don't attempt to make engines work with buggy clients. If a user has an existing 2.2 system, he suffers from this bug.

Currently, if without my patch, if application trigger empty "sg" with non-zero "nr", it will cause the client itself to crash before trigger RPC. Means the client cannot send related RPC to the server, so server is safe.

Here, the new flag "ORF_EMPTY_SGL" help the new client (with my patch) to avoid packing invalid information t to server, that will avoid server crash after fixing client side issue, in spite of the server or new. The flag "ORF_EMPTY_SGL" itself will not be packed on-wire, that avoids interoperability issue.

> if pack a "sgl_nr" for encoding which is 0 if "dcu_sgls == NULL", and for decoding unpack and check sgl_nr first, seems need not add ORF_EMPTY_SGL flag? (just to save one flag bit if needed).

We can NOT pack more information on-wire to indicate some special things, that will cause interoperability issue. We also cannot change the "nr" itself on-wire, it is used on server.

Reply:
OK...